pandas/core/strings.py

```diff
@@ -150,14 +150,20 @@ def _map_stringarray(
         an ndarray.
 
     """
-    from pandas.arrays import IntegerArray, StringArray
+    from pandas.arrays import IntegerArray, StringArray, BooleanArray
 
     mask = isna(arr)
 
     assert isinstance(arr, StringArray)
     arr = np.asarray(arr)
 
-    if is_integer_dtype(dtype):
+    if is_integer_dtype(dtype) or is_bool_dtype(dtype):
+        constructor: Union[Type[IntegerArray], Type[BooleanArray]]
+        if is_integer_dtype(dtype):
+            constructor = IntegerArray
+        else:
+            constructor = BooleanArray
```

Comment:
is there a way to combine the above if/else reading this is super confusing

Reply:
I don't see an easy way. `dtype.construct_array_type` isn't an option since we have a the functions using `na_map` use dtypes like `bool`, `int`, since they work with both object-dtype arrays returning numpy arrays, or StringArray returning EAs. So we can have either a NumPy dytpe or an extension type here.